Use default profile and switch country-specific profiles into a feature flag

melbeltagy opened the pull request:
Changes:
* Splitting each module's code into `org.niis.*` and `fi.dvv.*` packages for common and Finland-specific features respectively.
* Switch `application.properties` into `yaml` format for:
  * Clarity in properties structure
  * Better support with docker/k8s based environments and usage.
* Switching from `fi` profile to the new `xroad-catalog.country.fi.enabled` feature flag for both collector and lister modules
* Loading Finland-specific features based on the new `xroad-catalog.country.fi.enabled` feature flag.
* Using `application.yaml` as a common/base configuration file with all default configurations
* Using `dev` profile configuration for local development (to run service from IDE)
* Using `test` profile configuration for tests only
* Removing `sshtest` profile but keeping/updating the documentation
* Adding `application-fi.sample.yaml` and `application-production.sample.yaml` sample profiles
* Using persistence module only for entities and repositories by:
  * Splitting `services` from persistence module into either collector or lister modules
  * Persistence module no longer can run as an application
  * Persistence module no longer can be used to apply database changes as liquibase is already configured
* Grouping properties into different logical groups
* Removing unused properties
* Adding new `configuration` section in each module's `README.md` with all configurations related to the the module

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move common entities to org.niis...entity package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move common repositories to org.niis...repository package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move DTOs from persistence module to lister module (splitting into common/fi packages)

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move services from persistence module to collector module (splitting into common/fi packages)

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move lister's controllers into own (common/fi) controller package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move lister's endpoints into own (common/fi) endpoint package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move lister's services into own (common/fi) service package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move lister's converters into own (common/fi) converter package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move lister's interceptors into own common interceptor package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move lister's configurations into own common configuration package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move lister's exception into own common exception package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Split CatalogService into two (one for each lister/collector service) based on usage

Refs: XRDCAT-8

melbeltagy committed:
refactor: Split CompanyService into two (one for each lister/collector service) based on usage

Refs: XRDCAT-8

melbeltagy committed:
refactor: Split OrganizationService into two (one for each lister/collector service) based on usage

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix imports after refactorings

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move repositories' tests into corresponding (common/fi) service package

Refs: XRDCAT-8

melbeltagy committed:
refactor: Remove persistence' Application class and keep it for tests only

Refs: XRDCAT-8

melbeltagy committed:
fix: Add/fix tests initialization by scanning required packages

Refs: XRDCAT-8

melbeltagy committed:
chore: Cleanup/remove unused code and SQL files from persistence's tests

Refs: XRDCAT-8

melbeltagy committed:
refactor: Split collector's classes into default/vrk packages, add application configs to scan components in both packages

Refs: XRDCAT-8

melbeltagy committed:
refactor: Split collector's tests into default/vrk packages, fix newly broken ones, add application.properties

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move main classes to org.niis package

Refs: XRDCAT-8

melbeltagy committed:
chore: Remove redundant dependencies

Refs: XRDCAT-8

melbeltagy committed:
refactor: switch to application.properties

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix license

Refs: XRDCAT-8

melbeltagy committed:
refactor: Modify generation of classes from WSDL into package org.niis instead of fi.vrk

Refs: XRDCAT-8

melbeltagy committed:
refactor: Move ListerApplication into org.niis package

Refs: XRDCAT-8

melbeltagy committed:
refactor: combine all properties into application.properties (default profile)

Refs: XRDCAT-8

melbeltagy committed:
refactor: combine all properties into application.properties (default profile) and fi profile

Refs: XRDCAT-8

petkivim commented:
Since the Java package names are being updated, `fi.vrk.*` should be updated to `fi.dvv.*`. Nowadays, the Finnish Digital Agency uses the `dvv.fi` domain while `vrk.fi` is their old domain.

melbeltagy committed:
refactor: Split lister service's classes into (default/Finland) specific packages

Refs: XRDCAT-8

melbeltagy committed:
refactor: Undo moving repository tests to lister service (moved by mistake)

Refs: XRDCAT-8

melbeltagy committed:
chore: Clean-up unused DTO and corresponding tests

Refs: XRDCAT-8

melbeltagy committed:
refactor: split lister's tests into default/Finland packages and fix failing tests

Refs: XRDCAT-8

melbeltagy committed:
refactor: expose postgres port in docker compose (at port 4910) for usage with local development

Refs: XRDCAT-8

melbeltagy committed:
refactor: Update package from fi.vrk.* to fi.dvv.*.

Refs: XRDCAT-8

melbeltagy committed:
refactor: Update gradle's group's name to use org.niis prefix

Refs: XRDCAT-8

melbeltagy committed:
refactor: Rename *VrkConfiguration (classes' names) to *FinlandConfiguration

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix gradle build command for persistence (set build.gradle as a JAR only, not a spring boot application)

Refs: XRDCAT-8

melbeltagy committed:
chore: Organize imports

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix typo

Refs: XRDCAT-8

melbeltagy committed:
doc: Fix JavaDoc

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix dependency

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix failing pipeline (handle different timezone for some/failing tests)

Refs: XRDCAT-8

melbeltagy committed:
refactor: Unify checkstyle rules for main and test code, and upgrade checkstyle version

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix checkstyle issues

Refs: XRDCAT-8

melbeltagy committed:
fix: Switch to string block to fix checkstyle error with string operator

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix checkstyle issues

Refs: XRDCAT-8

melbeltagy committed:
chore: add application-dev.properties for both services with values for local development, remove passwords and value that must be configured for usage from application.properties

Refs: XRDCAT-8

melbeltagy committed:
chore: Increase log level for tests and local dev environments for easier debugging

Refs: XRDCAT-8

melbeltagy committed:
chore: Clean up SSH-Test profile related settings

Refs: XRDCAT-8

melbeltagy committed:
chore: Fix pipeline's failing tests (change open_api fetched date to 2nd of Jan to be in 2016 - something related to running pipeline in github on a diff timezone)

Refs: XRDCAT-8

melbeltagy committed:
chore: Fix pipeline's failing tests (change organization fetched date to 2nd of Jan to be in 2016 - something related to running pipeline in github on a diff timezone)

Refs: XRDCAT-8

melbeltagy committed:
chore: Fix pipeline's failing tests (change organization fetched date to 2nd of Jan to be in 2016 - something related to running pipeline in github on a diff timezone)

Refs: XRDCAT-8

melbeltagy committed:
chore: Fix pipeline's failing tests (change organization fetched date to 2nd of Jan to be in 2016 - something related to running pipeline in github on a diff timezone)

Refs: XRDCAT-8

melbeltagy committed:
refactor: Allow components scanning based on feature flag, and simplify usage of country specific configs

Refs: XRDCAT-8

melbeltagy committed:
refactor: Allow components scanning based on feature flag, and simplify usage of country specific configs

Refs: XRDCAT-8

melbeltagy committed:
refactor: Remove usage of "fi" profile in lister service (feature flag is in place)

Refs: XRDCAT-8

melbeltagy committed:
chore: add missing datasource.url for lister's dev profile

Refs: XRDCAT-8

melbeltagy committed:
chore: cleanup

Refs: XRDCAT-8

melbeltagy committed:
chore: don't use componse's name in services implementation

Refs: XRDCAT-8

melbeltagy committed:
refactor: rename main class to CollectorApplication

Refs: XRDCAT-8

melbeltagy committed:
refactor: Add a NewMembersEvent with listener and publisher components for usage in related tasks

Refs: XRDCAT-8

melbeltagy committed:
chore: don't use component's name in services implementation

Refs: XRDCAT-8

melbeltagy committed:
refactor: Refactor tasks initialization into using Spring components to use CDI, make default tasks independent of Finland queues by using NewMembersEvent

Refs: XRDCAT-8

melbeltagy committed:
chore: add license header to new classes

Refs: XRDCAT-8

melbeltagy committed:
chore: remove finland profile

Refs: XRDCAT-8

melbeltagy committed:
chore: add missing property for catalog service for local dev env

Refs: XRDCAT-8

melbeltagy committed:
refactor: convert properties files into yaml files (for better support with docker/k8s environments) and visual hierarchical structure for easy management

Refs: XRDCAT-8

melbeltagy committed:
chore: sort properties alphabetical in each group, and the groups into this order: spring, xroad-catalog, logging

Refs: XRDCAT-8

melbeltagy committed:
chore: cleanup unused properties

Refs: XRDCAT-8

melbeltagy committed:
chore: cleanup unused properties

Refs: XRDCAT-8

melbeltagy committed:
refactor: Load Finland specific properties into via its own FinlandTaskPoolConfiguration

Refs: XRDCAT-8

melbeltagy committed:
refactor: Group Finland specific properties into a nested properties under xroad-catalog.country.fi.*, and update docs

Refs: XRDCAT-8

melbeltagy committed:
refactor: Merge Finland's feature flag into Finland group

Refs: XRDCAT-8

melbeltagy committed:
chore: remove duplicate properties from application-dev.yaml

Refs: XRDCAT-8

melbeltagy committed:
refactor: Merge Finland's feature flag into Finland group

Refs: XRDCAT-8

melbeltagy committed:
refactor: Refactor properties into sub-hierarchy

Refs: XRDCAT-8

melbeltagy committed:
Merge branch 'refs/heads/develop' into XRDCAT-8-clearer-profiles

melbeltagy committed:
fix: Fix persistence tests (add fi feature flag)

Refs: XRDCAT-8

melbeltagy committed:
refactor: Convert to properties into yaml format

Refs: XRDCAT-8

melbeltagy committed:
chore: cleanup persistence

Refs: XRDCAT-8

melbeltagy committed:
fix: remove unused imports (fix checkstyle)

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix collector tests after recent refactorings

Refs: XRDCAT-8

melbeltagy committed:
fix: Fix checkstyle's issues

Refs: XRDCAT-8

melbeltagy committed:
refactor: Rename DevelopmentConfiguration to TestingConfiguration

Refs: XRDCAT-8

melbeltagy committed:
chore: cleanup unused exceptions from throws statement

Refs: XRDCAT-8

melbeltagy committed:
Merge branch 'refs/heads/develop' into XRDCAT-8-clearer-profiles

melbeltagy committed:
refactor: Use test profile in tests and remove redundant properties from test profile

Refs: XRDCAT-8

melbeltagy committed:
refactor: Use test profile in tests and remove redundant properties from test profile

Refs: XRDCAT-8

melbeltagy committed:
refactor: Use test profile in tests and remove redundant properties from test profile

Refs: XRDCAT-8

melbeltagy committed:
doc: update comment regarding fi enable/disable flag

Refs: XRDCAT-8

melbeltagy committed:
chore: clean and format application yaml files

Refs: XRDCAT-8

melbeltagy committed:
chore: order application properties into logical order (instead of alphabetical)

Refs: XRDCAT-8

melbeltagy committed:
chore: order application properties into logical order (instead of alphabetical)

Refs: XRDCAT-8

melbeltagy committed:
chore: remove unused liquibase dependency from lister service

Refs: XRDCAT-8

melbeltagy committed:
chore: Update docker compose, switch to application.yaml for compose, update Dockerfile to remove profiles settings, provide sample YAMLs

Refs: XRDCAT-8

melbeltagy committed:
docs: update README(s)

Refs: XRDCAT-8

melbeltagy committed:
chore: update yaml files

Refs: XRDCAT-8

melbeltagy committed:
docs: remove profiles related documentation from BUILD.md and replace by the feature flag

Refs: XRDCAT-8

melbeltagy committed:
docs: Add configurations section to collector's module README

Refs: XRDCAT-8

melbeltagy committed:
docs: Add configurations section to collector's module README

Refs: XRDCAT-8

melbeltagy committed:
chore: update application.yaml

Refs: XRDCAT-8

raits commented on BUILD.md:
```diff
@@ -41,48 +41,33 @@ as well), especially if you set up a local X-Road ecosystem.
+The flag is `xroad-catalog.country.fi.enabled` and it is set to `false` by default. When the flag is set to `true`,
+corresponding features will be enabled.
 
 X-Road Catalog supports adding new profiles. For example, new country-specific features should be added by creating a
```
Since we are now using feature flags instead of profiles, this text should also be update to reflect that.

raits commented on doc/xroad_catalog_installation_guide.md:
```diff
@@ -190,16 +190,18 @@ Optional parameters which can be configured in the same file are described below
 When using the `xroad-catalog-collector` module with the `FI` profile, the following additional optional parameters are
```
This should also mention the feature flag instead of the profile.

melbeltagy replied:
Nice catch.

raits commented on xroad-catalog-collector/README.md:
```diff
@@ -23,16 +23,96 @@ See also the [Installation Guide](../doc/xroad_catalog_installation_guide.md) an
+| [spring.jpa.open-in-view](https://docs.spring.io/spring-boot/appendix/application-properties/index.html#application-properties.data.spring.jpa.open-in-view)                         | N        | `false`                                          | Preferably `false` in production.                                                                                                                        ...
+| [spring.jpa.show-sql](https://docs.spring.io/spring-boot/appendix/application-properties/index.html#application-properties.data.spring.jpa.show-sql)                                 | N        | `false`                                          | Use to debug SQL statements executed. Preferably `false` in production.                                                                                  ...
+| [spring.jpa.hibernate.ddl-auto](https://docs.spring.io/spring-boot/appendix/application-properties/index.html#application-properties.data.spring.jpa.hibernate.ddl-auto)             | Y        | `none`                                           | Keep to `none` to avoid conflict with liquibase.                                                                                                         ...
+| spring.jpa.properties.hibernate.dialect                                                                                                                                              | Y        | `org.hibernate.dialect.PostgreSQLDialect`        | PostgreSQL is the supported RDMBS. But change to other dialect according to the used RDBMS, if needed. Available options can be found [here](https://doc ...
```
Since we only ship the PostgreSQL driver, I think we should stick to saying only PostgreSQL is supported for now.

raits commented on xroad-catalog-collector/README.md:
```diff
@@ -23,16 +23,96 @@ See also the [Installation Guide](../doc/xroad_catalog_installation_guide.md) an
+| `xroad-catalog.target.member-code`                     | Y        |                                              | A parameter for setting the X-Road member code.                                                                                                                                                                                                                                            ...
+| `xroad-catalog.ssl-keystore.location`                  | N        | `/etc/xroad/xroad-catalog/keystore`          | A parameter for setting the keystore's location that contains SSL certificates required to connect to X-Road host.                                                                                                                                                                         ...
+| `xroad-catalog.ssl-keystore.password`                  | N        | `changeit`                                   | A parameter for setting the keystore's password.                                                                                                                                                                                                                                           ...
+| `xroad-catalog.urls.security-server-host`              | Y        |                                              | A parameter for setting the security server host to connect to. e.g., http://security.server.host:5000                                                                                                                                                                                     ...
```
It would be better to use port 8080 here, as that is the default client port for the Security Server proxy.

raits commented on xroad-catalog-collector/src/main/java/fi/dvv/xroad/catalog/collector/configuration/FinlandTaskPoolConfiguration.java:
```diff
@@ -0,0 +1,68 @@
+@Configuration
+public class FinlandTaskPoolConfiguration {
+
+    // Parameters related to the "fi" profile
```
This should also mention the feature flag instead of profile.

raits commented on xroad-catalog-collector/src/main/java/fi/dvv/xroad/catalog/collector/tasks/FinlandTasksInitializer.java:
```diff
@@ -0,0 +1,67 @@
+    @SneakyThrows
+    public void onApplicationEvent(ApplicationStartedEvent ase) {
+
+        log.info("FI profile detected, starting up organizations and companies fetchers");
```
This should also be changed to mention the feature flag instead of profile.

raits commented on xroad-catalog-collector/src/test/java/org/niis/xroad/catalog/collector/tasks/ListClientsTaskTest.java:
```diff
@@ -84,31 +89,26 @@ public void testOnReceiveWhenFetchUnlimited() throws Exception {
             listClientsTask.run();
 
             verify(catalogService, times(1)).saveAllMembersAndSubsystems(any());
+            verify(newMembersEventPublisher, times(1)).publishNewMembersEvent(any());
```
I think we should also still verify that both members were passed in the event.

raits commented on xroad-catalog-collector/src/test/java/org/niis/xroad/catalog/collector/tasks/ListClientsTaskTest.java:
```diff
@@ -168,75 +162,59 @@ public void testOnReceiveWhenFetchNotUnlimitedButTimeIsInBetween() throws Except
 
             verify(catalogService, times(1)).saveAllMembersAndSubsystems(any());
-
+            verify(newMembersEventPublisher, times(1)).publishNewMembersEvent(any());
```
Same comment as above about checking the amount of members.

raits commented on xroad-catalog-lister/README.md:
```diff
@@ -10,6 +10,49 @@ A class diagram illustrating X-Road Catalog Lister implementation with the `defa
+| [spring.jpa.open-in-view](https://docs.spring.io/spring-boot/appendix/application-properties/index.html#application-properties.data.spring.jpa.open-in-view)                         | N        | `false`                                   | Preferably `false` in production.                                                                                                                               ...
+| [spring.jpa.show-sql](https://docs.spring.io/spring-boot/appendix/application-properties/index.html#application-properties.data.spring.jpa.show-sql)                                 | N        | `false`                                   | Use to debug SQL statements executed. Preferably `false` in production.                                                                                         ...
+| [spring.jpa.hibernate.ddl-auto](https://docs.spring.io/spring-boot/appendix/application-properties/index.html#application-properties.data.spring.jpa.hibernate.ddl-auto)             | Y        | `none`                                    | Keep to `none` to prevent hiberante from trying to update the database.                                                                                         ...
+| spring.jpa.properties.hibernate.dialect                                                                                                                                              | Y        | `org.hibernate.dialect.PostgreSQLDialect` | PostgreSQL is the supported RDMBS. But change to other dialect according to the used RDBMS, if needed. Available options can be found [here](https://docs.jboss ...
```
Here as well since we only ship the PostgreSQL driver I think we should say we only support it for now.

raits commented on xroad-catalog-persistence/src/test/java/fi/dvv/xroad/catalog/persistence/repository/OrganizationRepositoryTest.java:
```diff
@@ -74,7 +77,7 @@ public void testFindLatestFetched() {
         assertEquals(2016, latestFetched.getYear());
         assertEquals(Month.JANUARY, latestFetched.getMonth());
-        assertEquals(1, latestFetched.getDayOfMonth());
+        assertEquals(2, latestFetched.getDayOfMonth());
```
Why did this change?

melbeltagy replied:
The pipeline kept failing for some reason. So I had to change the day of the month in tests data in the `data.sql`.

raits commented on docker/collector/config/application.yaml:
I think this file should be called `application.sample.yaml` or something similar and renamed when used in the development docker setup. At the moment, if you change this you need to start discarding the changes from Git. It would probably also be a good idea to `.gitignore` this file to prevent it ending up in the index.

raits commented on docker/lister/config/application.yaml:
Same as above:

I think this file should be called `application.sample.yaml` or something similar and renamed when used in the development docker setup. At the moment, if you change this you need to start discarding the changes from Git. It would probably also be a good idea to `.gitignore` this file to prevent it ending up in the index.

raits commented on xroad-catalog-collector/README.md:
````diff
@@ -49,22 +129,27 @@ java -jar target/xroad-catalog-collector-1.0-SNAPSHOT.jar
 ## Run against a remote Security Server over an SSH tunnel
 
-First create an ssh tunnel to a local port:
+1. First create an ssh tunnel to a local port:
 
-```bash
-ssh -nNT -L <LOCAL_PORT>:<DESTINATION>:<DESTINATION_PORT> [USER@]SSH_SERVER
-```
+    ```bash
+    ssh -nNT -L <LOCAL_PORT>:<DESTINATION>:<DESTINATION_PORT> [USER@]SSH_SERVER
+    ```
 
-For example, there's a Security Server running on a machine `my-security-server.com` on an internal private network on
-port `80`. The Security Server is accessible from the machine `my-ssh-server.com`. To connect to the Security Server from
-the local machine using the local port `9000`, forward the connection using the following command:
+    For example, there's a Security Server running on a machine `my-security-server.com` on an internal private network on
+    port `80`. The Security Server is accessible from the machine `my-ssh-server.com`. To connect to the Security Server
+    from the local machine using the local port `9000`, forward the connection using the following command:
 
-```bash
-ssh -nNT -L 9000:my-security-server.com:80 my-user@my-ssh-server.com
-```
+    ```bash
+    ssh -nNT -L 9000:my-security-server.com:80 my-user@my-ssh-server.com
+    ```
 
-Then run the collector with profile `sshtest`:
+2. Then run the collector with profile `sshtest`:
 
-```bash
-java -Dspring.profiles.active=sshtest -jar build/libs/xroad-catalog-collector.jar --spring.config.name=collector,catalogdb
-```
+    ```bash
+    java -jar build/libs/xroad-catalog-collector.jar --xroad-catalog.security-server-host=http://localhost:<LOCAL_PORT>
+    ```
+    For our example and since the port is `9000`, the command would be:
+
+    ```bash
+    java -jar build/libs/xroad-catalog-collector.jar --xroad-catalog.security-server-host=http://localhost:9000
+    ```
````
Is this still applicable after removing the `sshtest` profile?

melbeltagy replied:
Yes. The JAR can still be used to connect to a remote security server via ssh.

raits commented on doc/xroad_catalog_installation_guide.md:
Parameter paths need to be updated due to changes in the TaskPoolConfiguration class.

raits commented on xroad-catalog-collector/src/main/java/org/niis/xroad/catalog/collector/configuration/CatalogDefaultConfiguration.java:
This class should probably be called `CollectorDefaultConfiguration` instead.

raits commented on xroad-catalog-collector/src/main/java/org/niis/xroad/catalog/collector/configuration/CatalogFinlandConfiguration.java:
This class should probably be called `CollectorFinlandConfiguration` instead.

raits commented on xroad-catalog-collector/src/main/java/org/niis/xroad/catalog/collector/tasks/ListClientsTask.java:
```diff
@@ -0,0 +1,127 @@
+
+            log.info("All subsystems ({}) sent to ListMethodsTask", subsystems.size());
+
+            newMembersEventPublisher.publishNewMembersEvent(newMembers.stream().map(Member::getMemberCode).toList());
```
What happens here if there is no listener registered for the event (which I guess is the case if `fi` is not enabled)?

I have a feeling this might lead to some internal queue growing until we run out of memory.

melbeltagy replied:
@raits, honestly, I am not sure.

Going through [the documentations](https://docs.spring.io/spring-framework/reference/core/beans/context-introduction.html#context-functionality-events), it says:
> but note that, by default, event listeners receive events synchronously. This means that the publishEvent() method blocks until all listeners have finished processing the event.

I also tried reading the Spring Framework's code base and I found this class [SimpleApplicationEventMulticaster](https://github.com/spring-projects/spring-framework/blob/main/spring-context/src/main/java/org/springframework/context/event/SimpleApplicationEventMulticaster.java#L142-L159). The highlighted lines shows that an Executor is being used and no queuing involved.   

I think that the behavior is based on instant notification and not a queue based one. They also mentioned somewhere that if a more complex solution is required, then Spring Integration should be used instead for a more effective Event based system.

What do you think?

raits replied:
Based on the code it does looks like a NOOP when there are no listeners registered. So it should probably be ok, but I think it might be good to profile this just in case an see if memory usage starts creeping up.

melbeltagy replied:
@raits, I created this simple task to test the memory usage of the collector service and made it to run every 250 milliseconds:
```java

public class TestingTask implements Runnable {
    @Autowired
    private final NewMembersEventPublisher newMembersEventPublisher;

    static final String MEM_NAME = "TestingTask %s";
    static int count = 1;

    @Override
    public void run() {
        log.info("TestingTask is running");
        newMembersEventPublisher.publishNewMembersEvent(Set.of(MEM_NAME.formatted(count++)));
    }
}
```

My findings are that when I profile the application, memory heap size increases by 4 MB every 38 seconds.  
When I stop recording and just monitor the memory usage, heap size increases by 4 MB, but almost every 1 minute.

Using the following event listener:
```java

public class TestEventListener {

    private List<String> newMembersCodes = new ArrayList<>();

    @EventListener
    public void handleNewMembersEvent(NewMembersEvent event) {
        log.info("New members detected: {} and are being handled by {}", event.newMembersCodes(), this);
        newMembersCodes.addAll(event.newMembersCodes());
    }
}
```


I tried enabling this listener, disabling it, adding the member codes to a list (as per the code above) to keep it in memory and affect the heapsize, or just printing it in logs without storing it in the list.  
All had the same memory and heap size. Nothing has changed at all.
I also kept the application going for half an hour to make sure, but no difference.

raits commented on xroad-catalog-collector/src/main/java/org/niis/xroad/catalog/collector/tasks/ListClientsTask.java:
```diff
@@ -0,0 +1,127 @@
+
+            log.info("All subsystems ({}) sent to ListMethodsTask", subsystems.size());
+
+            newMembersEventPublisher.publishNewMembersEvent(newMembers.stream().map(Member::getMemberCode).toList());
```
The previous implementation also used `toList()` here, but now looking at it I think it would make more sense to use a `Set` instead, since we don't want duplicates.

raits reviewed (changes requested):
Left some comments. Also, it would be good to check the rest of the documentation / codebase for references to profiles so that we avoid confusion.

And as a remark, in the future we should really consider splitting changes to the structure from changes to the functionality. Structure changes generate so much noise that it makes it difficult to navigate the actual changes to the functionality.

petkivim commented on xroad-catalog-lister/README.md:
```diff
@@ -10,6 +10,49 @@ A class diagram illustrating X-Road Catalog Lister implementation with the `defa
 
+## Configuration
+
+X-Road Catalog Lister can be configured by creating a copy of the [application.yaml](/src/main/resources/application.yaml) file.
```
The `application.yaml` file should should only contain properties that the user is expected to change, e.g., `spring.datasource.username`, `spring.datasource.password`, `xroad-catalog.shared-params-file`. Now the file contains properties that the user should not change, e.g., `spring.jpa.*`, `springdoc.packagesToScan`. It would be better if the default configuration was included in the application JAR file and the user could override the properties that need to be overridden using the `application.yaml` file. In that way, the user don't need to mind about the other properties. The same comment applies to the lister and collector modules.

melbeltagy replied:
I split the configurations section into different groups, `mandatory`, `optional`, and `fixed-values in application.yaml`. I hope this clarifies the configurations and meets your suggestion.

I left the `Required` column as some of these configurations parameters are conditionally mandatory. But it can be removed, if required.

melbeltagy replied:
@raits and @petkivim , since the latest structure of the document is now splitting the configurations into mandatory and optional sections, I removed the `Required` column from the READMEs

petkivim commented on xroad-catalog-lister/README.md:
```diff
@@ -10,6 +10,49 @@ A class diagram illustrating X-Road Catalog Lister implementation with the `defa
+X-Road Catalog Lister can be configured by creating a copy of the [application.yaml](/src/main/resources/application.yaml) file.
+The following properties are available:
+
+### Spring Boot Configurations
```
Consider removing properties that the user is not expected to change. For example, the application supports only PostgreSQL so there's no need tell how to change the dialect to support some other DB.

petkivim commented on BUILD.md:
```diff
@@ -41,48 +41,33 @@ as well), especially if you set up a local X-Road ecosystem.
+the `default` profile with all configurations in place.
 
-Multiple profiles can be activated at the same time by separating them with a comma, e.g., `spring.profiles.active=production,fi`.
+For Finland's national business registry, a feature flag is used to enable fetching additional data from the registry.
+The flag is `xroad-catalog.country.fi.enabled` and it is set to `false` by default. When the flag is set to `true`,
+corresponding features will be enabled.
```
Consider changing this to:

> Fetching additional data from the Finnish Business Registry and the Suomi.fi Finnish Service Catalogue can enabled using the `xroad-catalog.country.fi.enabled` flag. When the flag is set to `true`, fetching the data is enabled. The default value is `false`.

melbeltagy committed:
Merge branch 'refs/heads/develop' into XRDCAT-8-clearer-profiles

melbeltagy committed:
docs: update docs and comments to reference Finland feature flag instead of fi profile

melbeltagy committed:
docs: update docs to mention that postgre is the only supported RDBMS

Refs: XRDCAT-8

melbeltagy committed:
chore: fix names

Refs: XRDCAT-8

melbeltagy committed:
chore: publish list of new members' code as a set, instead of list

Refs: XRDCAT-8

melbeltagy committed:
chore: gitignore any file other than the sample yamls

Refs: XRDCAT-8

melbeltagy committed:
docs: update parameters paths

Refs: XRDCAT-8

melbeltagy committed:
docs: Update configurations sections in READMEs

Refs: XRDCAT-8

melbeltagy committed:
docs: Update configurations sections in READMEs

Refs: XRDCAT-8

raits commented on BUILD.md:
```diff
@@ -65,9 +65,7 @@ See [xroad-catalog-persistence/README.md](xroad-catalog-persistence/README.md#bu
 Profiles can be used to configure different features in X-Road Catalog. By default, X-Road Catalog supports 1 profile, 
 the `default` profile with all configurations in place.
```
I think since we only do have the one profile at the moment, this could probably be removed altogether.

raits commented on doc/xroad_catalog_installation_guide.md:
This file mentions the path `/etc/xroad/xroad-catalog/collector-production.properties`. This probably is not correct anymore. It will also change with the Docker configuration, but in the meanwhile we should use the up-to-date filename here and also in other places they are referenced.

raits commented on doc/xroad_catalog_installation_guide.md:
```diff
@@ -203,14 +203,14 @@ in effect:
 | `xroad-catalog.country.fi.fetch.external-update-after-days` | 7        | A parameter for setting the amount of days after which the X-Road Catalog Collector should consider Company and Organization data stale and try to fetch data from the external API again, e.g. value `7` means `after 7 days`.                                                                                                  |
 
-In addition, update the `xroad-catalog.shared-params-file` property value in `/etc/xroad/xroad-catalog/lister-production.properties`.
+In addition, update the `xroad-catalog.shared-params-file` property value.
```
I feel we should still leave a note here stating that this is for the lister configuration (as the previous ones are for the collector). It might in fact make sense to split it with a header. And then the passwords part keep as a common section below.

raits commented on xroad-catalog-collector/README.md:
```diff
@@ -29,90 +29,120 @@ X-Road Catalog Collector can be built by running:
+| `xroad-catalog.target.member-class`       | Y        |                                              | A parameter for setting the X-Road member class.                                                                                                                                                      | 1.0.0 |
+| `xroad-catalog.target.member-code`        | Y        |                                              | A parameter for setting the X-Road member code.                                                                                                                                                       | 1.0.0 |
+| `xroad-catalog.urls.security-server-host` | Y        |                                              | A parameter for setting the security server host to connect to. e.g., http://security.server.host:8080                                                                                                | 1.0.0 |
+| `xroad-catalog.urls.list-clients-host`    | Y        | `${xroad-catalog.urls.security-server-host}` | A parameter for setting the security server host to be used to get list of clients. Default value references `xroad-catalog.urls.security-server-host`.                                               | 1.0.0 |
+| `xroad-catalog.urls.webservices-endpoint` | Y        | `${xroad-catalog.urls.security-server-host}` | A parameter for setting the security server host to be used to get list of OpenAPI services, SOAP services, and list of services. Default value references `xroad-catalog.urls.security-server-host`. | 1.0.0 |
```
I think these aren't actually mandatory, since they default to the `xroad-catalog.urls.security-server-host`.

raits commented on xroad-catalog-collector/README.md:
```diff
@@ -29,90 +29,120 @@ X-Road Catalog Collector can be built by running:
+
+| Configurations                                     | Required | Defaults | Description                                                                                                                                                                   | Since |
+|----------------------------------------------------|----------|----------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------|
+| `xroad-catalog.country.fi.enabled`                 | Y        | `false`  | A parameter to enable/disable Finland specific features. If `false`, then all Finland specific features are disabled and all related parameters becomes irrelevant (not used) | 1.0.0 |
```
This also has a reasonable default, so I would say it is not required.

raits commented on xroad-catalog-lister/README.md:
```diff
@@ -12,32 +12,57 @@ See also the [Installation Guide](../doc/xroad_catalog_installation_guide.md) an
+
+| Configurations                     | Required | Defaults | Description                                                                                                           | Since |
+|------------------------------------|----------|----------|-----------------------------------------------------------------------------------------------------------------------|-------|
+| `xroad-catalog.country.fi.enabled` | Y        | `false`  | A parameter to enable/disable Finland specific features. If `false`, then all Finland specific endpoints are disabled | 1.0.0 |
```
Same here, I think it could be argued that this is not required as `false` is a reasonable default.

raits reviewed (changes requested):
A few more small things I noticed, but looking pretty good otherwise.

melbeltagy committed:
docs: Update configurations sections in documentations

Refs: XRDCAT-8

melbeltagy commented:
@raits , the configurations section in both READMEs for collector and lister services are updated as per your comments (moving reasonably defaulted values to proper optional sections).

And the configurations section in the installation guide is now updated by:
* Splitting collector from lister service
* Reference READMEs configurations sections (instead of repeating the configurations parameters list)
* Added specific instructions for how to enable Finland-specific features (in both services)

And also, added ToC for READMEs and installation guide to help clarifying how the documents are structured

raits commented on xroad-catalog-collector/README.md:
```diff
@@ -1,9 +1,31 @@
+
+## Table of Contents
+
+- [X-Road Catalog Collector](#x-road-catalog-collector)
+    - [Table of Contents](#table-of-contents)
+    - [Introduction to X-Road Catalog Collector](#introduction-to-x-road-catalog-collector)
+    - [Build](#build)
+    - [Configuration](#configuration)
+        - [Mandatory to Provide](#mandatory-to-provide)
+            - [Mandatory Configurations for Data Source and Liquibase](#mandatory-configurations-for-data-source-and-liquibase)
+            - [Mandatory Configurations for Common Features](#mandatory-configurations-for-common-features)
+            - [Mandatory Configurations for Finland-specific Features](#mandatory-configurations-for-finland-specific-features)
+        - [Optional configurations](#optional-configurations)
+            - [Optional Configurations for Common Features](#optional-configurations-for-common-features)
+            - [Optional Configurations for Finland-specific Features](#optional-configurations-for-finland-specific-features)
+        - [Fixed-Mandatory values to include in `application.yaml`](#fixed-mandatory-values-to-include-in-applicationyaml)
+            - [Fixed-Mandatory Values for Data Source and Liquibase](#fixed-mandatory-values-for-data-source-and-liquibase)
+            - [Fixed-Mandatory Values for Spring Boot Framework](#fixed-mandatory-values-for-spring-boot-framework)
+    - [Run](#run)
+    - [Run against a remote Security Server over an SSH tunnel](#run-against-a-remote-security-server-over-an-ssh-tunnel)
```
Since we use the asterix (`*`) style of unordered lists in other files, it would be good to use it here as well.

raits commented on xroad-catalog-lister/README.md:
```diff
@@ -1,6 +1,25 @@
+
+## Table of Contents
+
+- [X-Road Catalog Lister](#x-road-catalog-lister)
+    - [Table of Contents](#table-of-contents)
+    - [Introduction](#introduction)
+    - [Configuration](#configuration)
+        - [Mandatory to provide](#mandatory-to-provide)
+            - [Mandatory Configurations for Data Source and Liquibase](#mandatory-configurations-for-data-source-and-liquibase)
+            - [Mandatory Configurations for common Features](#mandatory-configurations-for-common-features)
+        - [Optional configurations](#optional-configurations)
+            - [Optional Configurations for Finland-specific Features](#optional-configurations-for-finland-specific-features)
+        - [Fixed-Mandatory values to include in `application.yaml`](#fixed-mandatory-values-to-include-in-applicationyaml)
+            - [Fixed-Mandatory Values for Data Source and Liquibase](#fixed-mandatory-values-for-data-source-and-liquibase)
+            - [Fixed-Mandatory values for OpenAPI documentation](#fixed-mandatory-values-for-openapi-documentation)
+    - [Build](#build)
+    - [Run](#run)
```
Same comment here about the list style.

raits reviewed (approved):
Looks good to me, just a minor comment about the list styles for the new ToC-s.

melbeltagy committed:
docs: Update list styles in READMEs

Refs: XRDCAT-8

melbeltagy committed:
docs: Remove "Required" column from configurations sections in READMEs

Refs: XRDCAT-8
